File: multipicker/src/main/java/com/edwardstock/multipicker/picker/views/FilesPresenter.java
package com.edwardstock.multipicker.picker.views;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;

import com.annimon.stream.Stream;
import com.edwardstock.multipicker.PickerConfig;
import com.edwardstock.multipicker.data.Dir;
import com.edwardstock.multipicker.data.MediaFile;
import com.edwardstock.multipicker.internal.MediaFileLoader;
import com.edwardstock.multipicker.picker.PickerConst;
import com.edwardstock.multipicker.picker.adapters.FilesAdapter;
import com.google.common.collect.ImmutableList;

import java.io.File;
import java.util.List;
import java.util.Locale;

import timber.log.Timber;

/**
 * android-multipicker. 2018
 * @author Eduard Maximovich [edward.vstock@gmail.com]
 */
public class FilesPresenter extends PickerPresenter<FilesView> implements MediaFileLoader.OnLoadListener {
    private final static String FILE_SCROLL_POSITION = "files_scroll_position";
    private PickerConfig mConfig;
    private FilesAdapter mAdapter;
    private Dir mDir;
    private int mSelectionCnt = 0;
    private int mScrollPosition = 0;

    public FilesPresenter() {
    }

    @Override
    public void updateFiles(MediaFileLoader loader) {
        callOnView(FilesView::showProgress);
        Timber.d("Updating files");
        loader.loadDeviceImages(mConfig, mDir, this);
    }

    public void handleExtras(Bundle bundle) {
        mDir = bundle.getParcelable(PickerConst.EXTRA_DIR);
        mConfig = bundle.getParcelable(PickerConst.EXTRA_CONFIG);
        mAdapter = new FilesAdapter(mConfig, null, this::onFileClick);
    }

    @Override
    public void onFilesLoadFailed(Throwable t) {
        callOnView(v -> {
            v.hideProgress();
            v.onError(t);
        });
    }

    @Override
    public void onFilesLoadedSuccess(List<MediaFile> images, List<Dir> dirList) {
        new Handler(Looper.getMainLooper()).post(() -> {
            callOnView(FilesView::hideProgress);
            callOnView(FilesView::hideRefreshProgress);

            final List<File> toDelete = Stream.of(images)
                    .map(MediaFile::getFile)
                    .filter(item -> !item.exists() || item.length() == 0)
                    .toList();

            new Thread(() -> {
                for (final File item : toDelete) {
                    callOnView(v -> {
                        v.removeFileFromMediaDB(item);
                    });
                }
            }).start();

            mAdapter.setData(Stream.of(images).filter(item -> item.exists() && item.length() > 0).toList());
            if (!images.isEmpty()) {
                mAdapter.notifyDataSetChanged();
            } else {
                callOnView(FilesView::showEmpty);
            }
        });
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (outState != null) {
            outState.putInt(FILE_SCROLL_POSITION, mScrollPosition);
        }
    }

    @Override
    public void onRestoreSavedState(Bundle savedState) {
        super.onRestoreSavedState(savedState);
        if (savedState != null && savedState.containsKey(FILE_SCROLL_POSITION)) {
            mScrollPosition = savedState.getInt(FILE_SCROLL_POSITION);
        }
    }

    @Override
    public void onViewAttach() {
        super.onViewAttach();
        if (mSelectionCnt == 0) {
            callOnView(v -> {
                v.setSelectionTitle("Выберите файл");
                v.setSelectionSubmitEnabled(false);
            });
        }
        Timber.d("Set adapter");
        callOnView(v -> {
            v.setOnRefreshListener(() -> {
                v.rescanFiles(() -> {
                    v.startUpdateFiles();
                    v.showRefreshProgress();
                });
            });
            v.setAdapter(mAdapter);
            v.scrollTo(mScrollPosition);
            v.setOnSelectionClearListener(v1 -> {
                v.clearSelection();
            });
            v.setOnSelectionDoneListener(v1 -> {
                v.finishWithResult();
            });
            v.setOnSelectionAllListener(v1 -> {
                v.selectAll();
            });


            Timber.d("Restore scroll: %d", mScrollPosition);
        });

    }

    public void setSelection(ImmutableList<MediaFile> selection) {
        if (mSelectionCnt == selection.size()) {
            return;
        }
        mSelectionCnt = selection.size();

        if (selection.isEmpty()) {
            callOnView(v -> {
                v.setSelectionTitle("Выберите файл");
                v.setSelectionSubmitEnabled(false);
            });
            return;
        }

        long numPhotos = Stream.of(selection).filter(item -> !item.isVideo()).count();
        long numVideos = selection.size() - numPhotos;

        if (numPhotos == 0 && numVideos == 0) {
            callOnView(v -> {
                v.setSelectionTitle("Выберите файл");
                v.setSelectionSubmitEnabled(false);
            });
            return;
        }

        if (numPhotos > 0 && numVideos == 0) {
            callOnView(v -> {
                v.setSelectionTitle(String.format(Locale.getDefault(), "Выбрано %d фото", numPhotos));
            });
        } else if (numPhotos == 0 && numVideos > 0) {
            callOnView(v -> {
                v.setSelectionTitle(String.format(Locale.getDefault(), "Выбрано %d видео", numVideos));
            });
        } else {
            callOnView(v -> {
                v.setSelectionTitle(String.format(Locale.getDefault(), "Выбрано %d фото и %d видео", numPhotos, numVideos));
            });
        }
        callOnView(v -> {
            v.setSelectionSubmitEnabled(true);
        });
    }

    public PickerConfig getConfig() {
        return mConfig;
    }

    public void setOnFileMeasuredListener(MediaFile file, FilesAdapter.OnMeasuredListener listener) {
        mAdapter.setOnFileMeasuredListener(file, listener);
    }

    public void setScroll(int position) {
        mScrollPosition = position;
    }

    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        callOnView(v -> {
            v.selectFile(data.getParcelableExtra(PickerConst.EXTRA_MEDIA_FILE));
        });
    }

    @Override
    protected void onViewDetach() {
        super.onViewDetach();
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        callOnView(v -> {
            v.setSelectionTitle("Выберите файл");
            v.setSelectionSubmitEnabled(false);
        });

    }

    private void onFileClick(MediaFile file, boolean isSelected, View sharedView) {
        callOnView(v -> {
            v.startPreview(file, sharedView);
        });
    }
}

File: multipicker/src/main/java/com/edwardstock/multipicker/picker/views/FilesPresenter.kt
package com.edwardstock.multipicker.picker.views

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import com.edwardstock.multipicker.PickerConfig
import com.edwardstock.multipicker.R
import com.edwardstock.multipicker.data.Dir
import com.edwardstock.multipicker.data.MediaFile
import com.edwardstock.multipicker.internal.MediaFileLoader
import com.edwardstock.multipicker.picker.PickerConst
import com.edwardstock.multipicker.picker.adapters.FilesAdapter
import timber.log.Timber
import java.io.File

/**
 * android-multipicker. 2018
 * @author Eduard Maximovich [edward.vstock@gmail.com]
 */
class FilesPresenter : BaseFsPresenter<FilesView>(), MediaFileLoader.OnLoadListener {
    private var mConfig: PickerConfig? = null
    private var mAdapter: FilesAdapter? = null
    private var mDir: Dir? = null
    private var mSelectionCnt = 0

    override fun updateFiles(loader: MediaFileLoader) {
        callOnView { obj: FilesView ->
            obj.showProgress()
            Timber.d("Updating files")
            loader.loadDeviceImages(mConfig!!, this)
        }

    }

    fun handleExtras(bundle: Bundle) {
        mDir = bundle.getParcelable(PickerConst.EXTRA_DIR)
        mConfig = bundle.getParcelable(PickerConst.EXTRA_CONFIG)
        mAdapter = FilesAdapter(mConfig!!, null) { file: MediaFile, isSelected: Boolean, sharedView: View ->
            onFileClick(file, isSelected, sharedView)
        }
    }

    override fun onFilesLoadFailed(t: Throwable?) {
        callOnView { v: FilesView ->
            v.hideProgress()
            v.onError(t)
        }
    }

    override fun onFilesLoadedSuccess(images: List<MediaFile>, dirList: List<Dir>) {
        Handler(Looper.getMainLooper()).post {
            callOnView { obj: FilesView -> obj.hideProgress() }
            var dirFiles: List<MediaFile>? = null
            for (d in dirList) {
                if (d.name == mDir?.name) {
                    dirFiles = d.files
                    break
                }
            }
            if (dirFiles == null || dirFiles.isEmpty()) {
                dirFiles = emptyList()
            }

            val toDelete = dirFiles
                    .filter { it.path != null }
                    .map { File(it.path!!) }
                    .filter { !it.exists() || it.length() == 0L }


            Thread {
                for (item in toDelete) {
                    callOnView { v: FilesView -> v.removeFileFromMediaDB(item) }
                }
            }.start()
            mAdapter!!.setData(dirFiles)
            if (dirFiles.isNotEmpty()) {
                mAdapter!!.notifyDataSetChanged()
            } else {
                callOnView { v: FilesView -> v.showEmpty() }
            }
        }
    }

    private var mScrollPosition = 0
    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState?.putInt(FILE_SCROLL_POSITION, mScrollPosition)
    }

    override fun onRestoreSavedState(savedState: Bundle?) {
        super.onRestoreSavedState(savedState)
        if (savedState != null && savedState.containsKey(FILE_SCROLL_POSITION)) {
            mScrollPosition = savedState.getInt(FILE_SCROLL_POSITION)
        }
    }

    public override fun onViewAttach() {
        super.onViewAttach()
        if (mSelectionCnt == 0) {
            callOnView { v: FilesView ->
                v.setSelectionTitle(R.string.mp_choose_file)
                v.setSelectionSubmitEnabled(false)
            }
        }
        Timber.d("Set adapter")
        callOnView { v: FilesView ->
            v.setAdapter(mAdapter!!)
            v.scrollTo(mScrollPosition)
            v.setOnSelectionClearListener { v1: View? -> v.clearSelection() }
            v.setOnSelectionDoneListener { v1: View? -> v.finishWithResult() }
            Timber.d("Restore scroll: %d", mScrollPosition)
        }
    }

    fun setSelection(selection: List<MediaFile>) {
        if (mSelectionCnt == selection.size) {
            return
        }
        mSelectionCnt = selection.size
        if (selection.isEmpty()) {
            callOnView { v: FilesView ->
                v.setSelectionTitle(R.string.mp_choose_file)
                v.setSelectionSubmitEnabled(false)
            }
            return
        }

        val numPhotos: Long = selection.filter { !it.isVideo }.count().toLong()
        val numVideos: Long = selection.size - numPhotos
        if (numPhotos == 0L && numVideos == 0L) {
            callOnView { v: FilesView ->
                v.setSelectionTitle(R.string.mp_choose_file)
                v.setSelectionSubmitEnabled(false)
            }
            return
        }
        if (numPhotos > 0 && numVideos == 0L) {
            callOnView {
                it.setSelectionTitleN(R.plurals.mp_choosed_n_photos, numPhotos.toInt())
            }
        } else if (numPhotos == 0L && numVideos > 0) {
            callOnView {
                it.setSelectionTitleN(R.plurals.mp_choosed_n_videos, numVideos.toInt())
            }
        } else {
            callOnView {
                it.setSelectionTitlePhotosAndVideos(
                        R.plurals.mp_choosed_n_photos_and_videos_prefix, numPhotos.toInt(),
                        R.plurals.mp_choosed_n_photos_and_videos_suffix, numVideos.toInt()
                )
            }
        }
        callOnView { v: FilesView -> v.setSelectionSubmitEnabled(true) }
    }

    val config: PickerConfig?
        get() = mConfig

    fun setOnFileMeasuredListener(file: MediaFile?, listener: (file: MediaFile) -> Unit) {
        mAdapter!!.setOnFileMeasuredListener(file, listener)
    }

    fun setScroll(position: Int) {
        mScrollPosition = position
    }

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        callOnView { v: FilesView ->
            v.setSelectionTitle(R.string.mp_choose_file)
            v.setSelectionSubmitEnabled(false)
        }
    }

    @Suppress("UNUSED_PARAMETER")
    private fun onFileClick(file: MediaFile, isSelected: Boolean, sharedView: View) {
        callOnView { v: FilesView -> v.startPreview(file, sharedView) }
    }

    companion object {
        private const val FILE_SCROLL_POSITION = "files_scroll_position"
    }
}